cmake: Add option to fail configuration instead of auto-reconfiguration. And make it default

code-of-kpp opened the pull request:
Qrator Labs s.r.o. contributes this under the terms of the Apache-2.0 license

This reverts commit 89419ceb9bd60121a029430b26aa4506c517f01d, turns `FAIL_ON_UNSUPPORTED_OPTIONS_COMBINATION` `ON` and provides few fixes for unbundled build

Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add a CMake option to fail configuration instead of auto-reconfiguration, enabled by default.


- В CMake добавлена опция FAIL_ON_UNSUPPORTED_OPTIONS_COMBINATION (включена по-умолчанию). Теперь если указать несовместимые параметры CMake, например `-DENABLE_FOO=OFF` и `-DUSE_INTERNAL_FOO_LIBRARY=ON`, CMake конфигурация завершится с ошибкой
- Устранены некоторые проблемы при сбокре unbundled - в CI unbundled версия собирается с MySQL и ODBC.
- Arrow и LLVM в unbundled версии в CI теперь собираются с использованием пакетов из репозиториев разработчиков этих библиотек
- Поскольку poco-clickhouse и upstream poco стали слишком сильно отличаться друг от друга, contrib-версия poco теперь используется по умолчанию всегда, даже в unbundled сборке.
- Некоторые библиотеки, недоступные в репозиториях debian/ubuntu и не имеющие официального репозитория разработчиков, теперь не отключаются по-умолчанию в unbundled сборке. Вместо этого будет испльзована contrib версия.

Detailed description / Documentation draft:

Предлагаю добавить в документацию для разработчиков:

Типичные ошибки при сборке, что они означают, как лечить.

пункт N.1:  линкер не может найти библиотеку `-lfoo` на самых последних этапах компиляции/сборки. 
что означает: где-то в cmake файлах написано `target_link_libraries(something .... foo)` подразуменвается, что foo должна быть target (возможно, imported), но его забыли определить или определили в неправильном scope. В результате cmake считает, что `foo`  - это строка и говорит линкеру искать библиотеку libfoo.a, например: `-lfoo`.
что делать: найти директиву `target_link_libraries(... foo)` и добавить где-то перед ней проверку `if(NOT TARGET foo) message(FATAL_ERROR ...`. Запустить cmake, убедиться, что теперь до компиляции не до ходит.  Дальше найти все места, где определна `foo` (наверняка через add_library) и проверить, что они выполняются в правильно порядке (до `target_link_libraries(... foo)`) и в правильном scope (например, imported targets должны добавляться в top-level scope: в `cmake/find/foo.cmake`, который вызывается из главного CMakeLists.txt. Например, если imported target `foo` добавляется в contrib/CMakeLists.txt через `add_subdirectory()`, то он не будет доступен в других `add_subdirectory()` в том же contrib/CMakeLists.txt.

code-of-kpp committed:
cmake: message when ThinLTO is enabled without clang

Squashed:
- ENABLE_THINLTO shouldn't be enabled by default when tests or sanitizers are enabled

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: enable FAIL_ON_UNSUPPORTED_OPTIONS_COMBINATION by default

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

akuzm commented:
Somehow it fails to build for fast test. It uses a special configuration with limited submodules.
```
CMake Error at cmake/find/termcap.cmake:9 (message):
  Can't find tinfo, ncurses or termcap library
Call Stack (most recent call first):
  CMakeLists.txt:387 (include)


2020-08-14 02:23:40 -- Configuring incomplete, errors occurred!
```
The error is in this log: https://clickhouse-test-reports.s3.yandex.net/13687/7459ec61079b78948c195f9115012e9233ebb4a6/fast_test/runlog.out.log
There is a link to it in Fast Test check details.

Not sure what is the best way to reproduce, maybe run it locally:
```
PULL_REQUEST_NUMBER=13687 COMMIT_SHA=7459ec61079b78948c195f9115012e9233ebb4a6 docker/test/fasttest/run.sh
```

I'll add `force tests` tag for now, so that fast test is ignored and other checks are ran.

podshumok commented:
@akuzm: thanks!

code-of-kpp committed:
cmake: don't enable internal llvm with disabled embedded compiler

Even if UNBUNDLED is off

Squashed:
- cmake: fix misleading error message about internal llvm on unsupported arches
- cmake: re-do some checks for LLVM options

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: make sure ENABLE_GRPC is off by default when PROTOBUF is OFF by default

Squashed:
- cmake: make sure protobuf compiler and snappy are there for unbundled build
- cmake: Turn ON internal GRPC library by default until relevant features will be implemented
- cmake: allow unbundled grpc

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: disable jemalloc on alternative arches and with sanitizers

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
Revert "Revert "cmake: Add option to fail configuration instead of auto-reconfiguration""

This reverts commit 89419ceb9bd60121a029430b26aa4506c517f01d
(curl part moved to separate commit)

Squashed:
- termcap removed
- fix for cassandra on apple merged
- cmake: fix "Can't find system zlib library" in unbundled build
- cmake: fix condition when testing for rdkafka platform
- cmake: PROTOBUF_OLD_ABI_COMPAT shouldn't be enabled for internal protobuf

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: set ENABLE_ORC to OFF by default on platforms that don't support protobuf or parquet

Squashed:
- cmake: Better diagnostic message about broken ORC dependencies

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: ARM: make sure cpuid is not enabled by default

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: unbundled + use_static_libs = no ldap

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: set USE_INTERNAL_SPARSEHASH_LIBRARY=ON by default

as we don't know any distribution with sparsehash-c11 package

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: USE_INTERNAL_H3_LIBRARY is always ON by default

as I don't know any system with package for it

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: disable external avro as we don't support it yet

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: fix parquet/arrow variables

squashed:

- sync FindArrow.cmake and FindParquet.cmake with arrow repo
- unbundled: add arrow to dbms link libraries
- cmake: fix-up unbundled Arrow support

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: always use internal poco library as it went too far from upstream

Squashed:
- cmake: move USE_INTERNAL_POCO_LIBRARY option to separate file
- point contrib/poco to "add <atomic> to TCPServerDispatcher"

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: disable internal ORC when using external Parquet

Squashed:
- fix internal poco linking with `zlib`

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: contrib/curl-cmake: use FAIL_ON_UNSUPPORTED_OPTIONS_COMBINATION

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
docker/packager/deb: new dependencies for unbundled

Squashed:
- docker/packager/deb: add arrow repo
- docker/packager/deb: cleanup
- docker/packager/deb: Add thrift
- docker/packager/deb: add myariadbclient
- docker/packager/deb: remove poco-dev as we always use contrib/ version now
- docker/packager/deb: add libssl-dev (unbundled) and restructure
- docker/packager/deb: fix repo url
- docker/packager/deb: fix tzdata not updated in build.sh
- docker/packager/deb: add mysqlclient-dev

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
Revert "Merge pull request #13689 from ClickHouse/remove-termcap"

This reverts commit da8d8f35ddcbd7071ce6ae1e521e447aa4a169d2, reversing
changes made to c3d0a87b8bf80665dbf529f2d18a6c847934ed7d.

code-of-kpp committed:
unbundled: link with termcap when using static llvm

ld.lld: error: undefined symbol: setupterm
referenced by Process.cpp.o:(llvm::sys::Process::FileDescriptorHasColors(int)) in archive /usr/lib/llvm-9/lib/libLLVMSupport.a

Squashed:
- fixup termcap + external llvm

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
unbundled: add thrift when using external parquet

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
test builds: docker/packager/packager: re-enable mysql

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: USE_INTERNAL_REPLXX_LIBRARY is ON by default now

As we don't now any system that ships it as a package

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
re-enable replxx in packager for unbundled

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
unbundled: fix detection of mariadb headers

Squashed:
- unbundled: add warning about mariadbclient dependencies

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: find/s3: use FAIL_ON_UNSUPPORTED_OPTIONS_COMBINATION

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: set USE_INTERNAL_ORC and USE_INTERNAL_HDFS3 to OFF by default even in unbundled

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: USE_INTERNAL_GSASL_LIBRARY by default when USE_STATIC_LIBRARIES

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: unbundled: libxml2 requires liblzma when building static

Squashed:
- fixup hdfs3-libxml2

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: unbundled protobuf: don't expect old c++ API by default

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: Fix odbc in unbudled

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: use bundled rdkafaka for in unbundled tests for now

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: fix potential curl linking problems

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
cmake: when USE_STATIC_LIBRARIES prefer bundled brotli

even if UNBUNDLED=ON because libbrotli*.a and similar are not
available in most of the systems

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

akuzm commented:
https://proxy.sandbox.yandex-team.ru/1678184909/test_run.txt.out.log

The PVS studio failure is related to the changes. The docker command to run it is like this:
```
docker run --volume=<full-path-to-clickhouse-repository>:/repo_folder --volume=<full-path-to-some-empty-dir-for-test-output>:/test_output -e LICENCE_NAME=ClickHouse -e LICENCE_KEY=<vault record #4726> -e CC=gcc-9 -e CXX=g++-9 yandex/clickhouse-pvs-test:latest
```

You don't have the license key, so it probably won' work as is. But seems like it just does `cmake .` without any parameters ([see the source here](https://github.com/ClickHouse/ClickHouse/blob/d806e0c0522978cf7fc1b06d6ebf020c1311079e/docker/test/pvs/Dockerfile#L29)) and it fails, could you please try it locally?

akuzm commented:
`01396_inactive_replica_cleanup_nodes` is a known flaky test

code-of-kpp committed:
dockerfiles: verify 3rd party apt repos, disable LLVM in PVS check

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

code-of-kpp committed:
sync unbundled/build.sh with a0ae1b2b42950df47a677525621047f2eb9eeaec

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

akuzm commented:
The container for PVS studio didn't rebuild, because it wasn't in the list. A fix is here: https://github.com/ClickHouse/ClickHouse/pull/14052
You can copy it to this one if you wish.

Do you know what's wrong with the unbundled build container? It complains about not being able to find libparquet-dev: https://clickhouse-test-reports.s3.yandex.net/13687/60be6b21906104aaad071b9d978575ffbe0f4493/push_to_dockerhub/dockerbuild.out.33.log

podshumok commented:
> Do you know what's wrong with the unbundled build container?

Yes, I think I do. It uses old base image version instead of the new one, provided in this PR. I indicated the dependency in `images.json`, but looks like it doesn't work with base images - they don't get overridden. I'm not sure how to fix this.I would start with using docker `ARG` statements, i.g. something like
```Dockerfile

ARG BASE_IMAGE=yandex/fancy-clickhouse-image-builder:latest

FROM ${BASE_IMAGE}
``` 
And then `docker build . -t yandex/dependent-image  --build-arg=yandex/fancy-clickhouse-image-builder:PR13687`

podshumok commented:
> A fix is here: #14052
> You can copy it to this one if you wish.

I did, I think it would be easier

akuzm commented:
>It uses old base image version instead of the new one, provided in this PR. I indicated the dependency in `images.json`, but looks like it doesn't work with base images - they don't get overridden.

To work around this, is it possible to merge the base image changes first, in a separate PR? I asked my colleagues to look into the docker dependency problem, but not sure when they'll fix it.

code-of-kpp committed:
REVERSE ME: workaround for docker/packager/unbundled

Please, reverse this commit after docker/packager/deb image update

Signed-off-by: Konstantin Podshumok <kpp.live+signed@gmail.com>

podshumok commented:
let's test another workaround attempt I just committed. If it fails, we'll have to split the PR, I guess...

bharatnc commented:
This is causing multiple errors when I build project using cmake:

1. ccache which I was able to remediate by installing it `apt install ccache`
2. `ThinLTO`. This check fails:

```
        set (CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO "${CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO} -flto=thin")
    elseif (ENABLE_THINLTO)
        message (${RECONFIGURE_MESSAGE_LEVEL} "Cannot enable ThinLTO")
```
can only make the cmake project reload work by commenting out the `    elseif (ENABLE_THINLTO)` block under [ClickHouse/CMakeLists.txt ](https://github.com/ClickHouse/ClickHouse/blob/master/CMakeLists.txt#L281) file.

podshumok commented:
We are discussing about CCache - and it will probably be disabled by default on systems with no ccache installed. Currently you can overcome this with `-D"ENABLE_CCACHE"=OFF` (or by installing ccache - which will save you some time probably).

With `ENABLE_THINLTO` error is completely legit. One way or another you are trying to enable it while it can't be and wouldn't done. If you need to enable it, make sure you are using clang, tests and sanitizers are turned off.

Anyway, you can enforce old behavior by calling CMake with `-D"FAIL_ON_UNSUPPORTED_OPTIONS_COMBINATION"=OFF`)

bharatnc commented:
> We are discussing about CCache - and it will probably be disabled by default on systems with no ccache installed. Currently you can overcome this with `-D"ENABLE_CCACHE"=OFF` (or by installing ccache - which will save you some time probably).
> 
> With `ENABLE_THINLTO` error is completely legit. One way or another you are trying to enable it while it can't be and wouldn't done. If you need to enable it, make sure you are using clang, tests and sanitizers are turned off.
> 
> Anyway, you can enforce old behavior by calling CMake with `-D"FAIL_ON_UNSUPPORTED_OPTIONS_COMBINATION"=OFF`)

sure, sounds good, thanks for the explanation :+1: .

azat commented on cmake/find/poco.cmake:
```diff
@@ -0,0 +1,10 @@
+option (USE_INTERNAL_POCO_LIBRARY "Use internal Poco library" ON)
```
Looks like it is better to add a comment why it does not use `NOT_UNBUNDLED` as a default value (otherwise someone like me, may make an attempt to change this).

Something like

```cmake
# unbundled poco has too much issues, and it is not recommended to use it even for unbundled build
# Here is the list of what may not work with unbundled poco:
# - s3 support will not work (since it requires modified Poco::Net::HTTPClientSession::setProxy())
# - custom http headers will not work
# - some build may fails (due to some patches were not accepted into upstream)
```
